### PR TITLE
Package handling (tarball and zip)

### PR DESCRIPTION
**Pull request by pirtnea**
Including edits to directory structure, applying tmc-langs for parsing code, and refactoring. :put_litter_in_its_place:

### CODE REVIEW

File: app/jobs/exercise_verifier_job.rb [hunk 1]
```diff
@@ -20,28 +20,21 @@ def perform(exercise)
...
-    `cd Stub/ && tar -cpf ../StubPackage.tar * && cd ..`
-  end
-
-  def create_model_solution_tar(exercise)
-    exercise.create_file('model_solution_file')
-    exercise.create_file('testfile')
-
-    `cd ModelSolution/ && tar -cpf ../ModelSolutionPackage.tar * && cd ..`
+    `cd ./submission_generation/tmp/Submission_#{exercise.id}/model/ && tar -cpf ../../../packages/ModelSolutionPackage_#{exercise.id}.tar * && cd ../../../..`
+    `cd ./submission_generation/tmp/Submission_#{exercise.id}/stub/ && tar -cpf ../../../packages/StubPackage_#{exercise.id}.tar * && cd ../../../..`
```
Comment (nygrenh) [inline app/jobs/exercise_verifier_job.rb]: Why are you changing directories here? Can't you just include the relative path to the input files in the arguments?

File: app/jobs/exercise_verifier_job.rb [hunk 2]
```diff
@@ -51,7 +44,7 @@ def send_package_to_sandbox(message, progress, exercise, token, package_name)
 
     token == 'KISSA_STUB' ? exercise.testing_stub! : exercise.testing_model_solution!
 
-    File.open(package_name, 'r') do |tar_file|
+    File.open('./submission_generation/packages/' + package_name, 'r') do |tar_file|
```
Comment (nygrenh) [inline app/jobs/exercise_verifier_job.rb]: `Rails.root.join('submission_generation', 'packages', package_name)`

File: app/jobs/exercise_verifier_job.rb [hunk 3]
```diff
@@ -60,6 +53,9 @@ def send_package_to_sandbox(message, progress, exercise, token, package_name)
 
   def sandbox_post(tar_file, exercise, token)
     response = RestClient.post post_url, file: tar_file, notify: results_url(exercise), token: token
+
+    `rm #{tar_file.path}`
```
Comment (nygrenh) [inline app/jobs/exercise_verifier_job.rb]: What if sandbox mysteriously dies and never sends a response?

File: app/models/exercise.rb [hunk 4]
```diff
@@ -12,25 +15,25 @@ class Exercise < ApplicationRecord
 
   enum status: %i[status_undefined saved testing_stub testing_model_solution finished error]
 
-  def create_file(file_type)
-    self_code = code
+  def create_submission
+    # muisto: self.code = code.gsub(%r{\/\/\sBEGIN SOLUTION\n(.*?\n)*\/\/\sEND SOLUTION}, '')
```
Comment (nygrenh) [inline app/models/exercise.rb]: Some memories are best left forgotten.

File: app/models/exercise.rb [hunk 5]
```diff
@@ -12,25 +15,25 @@ class Exercise < ApplicationRecord
...
 
-    if file_type == 'stubfile'
-      self.code = code.gsub(%r{\/\/\sBEGIN SOLUTION\n(.*?\n)*\/\/\sEND SOLUTION}, '')
-      write_to_file('Stub/src/Stub.java', MainClassGenerator.new, 'Stub')
+    if !Dir.exist?("submission_generation/tmp/Submission_#{id}/")
```
Comment (nygrenh) [inline app/models/exercise.rb]: Use `Rails.root.join` for constructing paths. Same thing below.

File: app/models/exercise.rb [hunk 6]
```diff
@@ -63,12 +66,12 @@ def generate_data_for_frontend(sandbox_status, test_output, token)
...
     sandbox_results[:passed] = false unless test_output['status'] == 'PASSED'
 
-    # Update exercises sandbox_results
+    # Update exercise's sandbox_results
     save!
   end
 
   def exercise_errors(test_output, token)
-    test_results(test_output)
+    test_results(test_output) if token == 'MODEL_KISSA'
```
Comment (nygrenh) [inline app/models/exercise.rb]: `MODEL_KISSA`?

File: bin/pre_setup [hunk 7]
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
+git submodule update --init --recursive
+cd ext/tmc-langs
+mvn package -DskipTests=true
```
Comment (nygrenh) [inline bin/pre_setup]: No newline at the end of file.

File: spec/jobs/exercise_verifier_job_spec.rb [hunk 8]
```diff
@@ -7,6 +7,8 @@
     let(:exercise) { FactoryGirl.create(:exercise) }
     before :each do
       allow_any_instance_of(ExerciseVerifierJob).to receive(:sandbox_post)
+      # allow_any_instance_of(ExerciseVerifierJob).to receive(:create_tar_files)
+      # allow_any_instance_of(Exercise).to receive(:create_model_solution_and_stub)
```
Comment (nygrenh) [inline spec/jobs/exercise_verifier_job_spec.rb]: Dead code.

File: app/jobs/exercise_verifier_job.rb [hunk 9]
```diff
@@ -17,41 +17,58 @@ def perform(exercise)
...
+    end
   end
 
-  def create_stub_tar(exercise)
-    exercise.create_file('stubfile')
-
-    `cd Stub/ && tar -cpf ../StubPackage.tar * && cd ..`
+  def exercise_modified?(exercise)
+    if Dir.exist?(Rails.root.join('submission_generation', 'packages', "assignment_#{exercise.assignment.id}")) &&
+       Dir.exist?(Rails.root.join('submission_generation', 'packages', "assignment_#{exercise.assignment.id}", "exercise_#{exercise.id}"))
```
Comment (nygrenh) [inline app/jobs/exercise_verifier_job.rb]: Maybe extract `Rails.root.join('submission_generation', 'packages', "assignment_#{exercise.assignment.id}"` into its own method?

File: app/jobs/exercise_verifier_job.rb [hunk 10]
```diff
@@ -17,41 +17,58 @@ def perform(exercise)
...
   end
 
-  def create_model_solution_tar(exercise)
-    exercise.create_file('model_solution_file')
-    exercise.create_file('testfile')
+  def create_tar_files(exercise)
+    exercise.create_submission
 
-    `cd ModelSolution/ && tar -cpf ../ModelSolutionPackage.tar * && cd ..`
+    `cd #{Rails.root.join('submission_generation', 'tmp', "Submission_#{exercise.id}", 'model').to_s} && tar -cpf \
```
Comment (nygrenh) [inline app/jobs/exercise_verifier_job.rb]: Can you do this without changing the working directory?

File: app/jobs/exercise_verifier_job.rb (hunk 3 above)
Comment (pirtnea) [inline app/jobs/exercise_verifier_job.rb]: Yes

**Pull request merged**